packages/bitcore-node/src/modules/ripple/api/csp.ts

```diff
@@ -0,0 +1,140 @@
+import { CSP } from '../../../types/namespaces/ChainStateProvider';
+import { InternalStateProvider } from '../../../providers/chain-state/internal/internal';
+import { RippleAPI } from 'ripple-lib';
+import { Readable } from 'stream';
+import { Storage } from '../../../services/storage';
+import { ChainNetwork } from '../../../types/ChainNetwork';
+import Config from '../../../config';
+import { FormattedTransactionType } from 'ripple-lib/dist/npm/transaction/types';
+
+export class RippleStateProvider extends InternalStateProvider implements CSP.IChainStateService {
+  config: any;
+  static client?: RippleAPI;
+
+  constructor(public chain: string = 'XRP') {
+    super(chain);
+    this.config = Config.chains[this.chain];
+  }
+
+  async getClient(network: string) {
+    try {
+      if (RippleStateProvider.client) {
+        await RippleStateProvider.client.getLedger();
+      }
+    } catch (e) {
+      RippleStateProvider.client = undefined;
+    }
+    if (!RippleStateProvider.client) {
+      const networkConfig = this.config[network];
+      const provider = networkConfig.provider;
+      const host = provider.host || 'localhost';
+      const protocol = provider.protocol || 'wss';
+      const portString = provider.port;
+      const connUrl = portString ? `${protocol}://${host}:${portString}` : `${protocol}://${host}`;
+      RippleStateProvider.client = new RippleAPI({ server: connUrl });
+      await RippleStateProvider.client.connect();
+    }
+    return RippleStateProvider.client;
+  }
+
+  async getBalanceForAddress(params: CSP.GetBalanceForAddressParams) {
+    const client = await this.getClient(params.network);
+    const info = await client.getAccountInfo(params.address);
+    const confirmed = Number(info.xrpBalance);
+    const balance = confirmed;
+    const unconfirmed = 0;
+    return { confirmed, unconfirmed, balance };
+  }
+
+  async getBlock(params: CSP.GetBlockParams) {
+    const client = await this.getClient(params.network);
+    const ledger = await client.getLedger({ includeTransactions: true, ledgerHash: params.blockId });
+    return ledger as any;
+  }
+
+  async getFee(params: CSP.GetEstimateSmartFeeParams) {
+    const client = await this.getClient(params.network);
+    const fee = await client.getFee();
+    return fee;
+  }
+
+  async broadcastTransaction(params: CSP.BroadcastTransactionParams) {
+    const client = await this.getClient(params.network);
+    return client.submit(params.rawTx);
+  }
+
+  async getWalletBalance(params: CSP.GetWalletBalanceParams) {
+    const { chain, network } = params;
+    const addresses = await this.getWalletAddresses(params.wallet._id!);
```

Comment:
Do XRP wallets have more that one address?

Reply:
It could, for instance for accounting we may want to pull the transaction history of multiple addresses. Bitcore-node treats each wallet as a bucket of addresses, even if it only has one address